rmw_fastrtps_shared_cpp/src/rmw_node_info_and_types.cpp

```diff
@@ -0,0 +1,490 @@
+    return ret;
+  }
+
+  // Get participant pointer from node
+  if (node->implementation_identifier != identifier) {
+    RMW_SET_ERROR_MSG("node handle not from this implementation");
+    return RMW_RET_ERROR;
+  }
+  return RMW_RET_OK;
+}
+
+/**
+ * Access the slave Listeners, which are the ones that have the topicnamesandtypes member
+ * Get info from publisher and subscriber
+ * Combined results from the two lists
+ *
+ * @param topic_cache cache with topic information
+ * @param topics [out] resulting topics
+ * @param node_guid_ to find information for
+ * @param no_demangle true if demangling will not occur
+ */
+void
+__accumulate_topics(
+  const LockedObject<TopicCache> & topic_cache,
+  std::map<std::string, std::set<std::string>> & topics,
+  const GUID_t & node_guid_,
+  bool no_demangle)
+{
+  std::lock_guard<std::mutex> guard(topic_cache.getMutex());
+  const auto & node_topics = topic_cache.getParticipantToTopics().find(node_guid_);
+  if (node_topics == topic_cache.getParticipantToTopics().end()) {
+    RCUTILS_LOG_DEBUG_NAMED(
+      kLoggerTag,
+      "No topics found for node");
+    return;
+  }
+  for (auto & topic_pair : node_topics->second) {
+    if (!no_demangle && _get_ros_prefix_if_exists(topic_pair.first) != ros_topic_prefix) {
+      // if we are demangling and this is not prefixed with rt/, skip it
+      continue;
+    }
+    RCUTILS_LOG_DEBUG_NAMED(
+      kLoggerTag,
+      "accumulate_topics: Found topic %s",
+      topic_pair.first.c_str());
+
+    topics[topic_pair.first].insert(topic_pair.second.begin(),
+      topic_pair.second.end());
+  }
+}
+
+/**
+ * Copy topic data to results
+ *
+ * @param topics to copy over
+ * @param allocator to use
+ * @param no_demangle true if demangling will not occur
+ * @param topic_names_and_types [out] final rmw result
+ * @return RMW_RET_OK if successful
+ */
+rmw_ret_t
+__copy_data_to_results(
+  const std::map<std::string, std::set<std::string>> & topics,
+  rcutils_allocator_t * allocator,
+  bool no_demangle,
+  rmw_names_and_types_t * topic_names_and_types)
+{
+  // Copy data to results handle
+  if (topics.size() > 0) {
+    // Setup string array to store names
+    rmw_ret_t rmw_ret = rmw_names_and_types_init(topic_names_and_types, topics.size(), allocator);
+    if (rmw_ret != RMW_RET_OK) {
+      return rmw_ret;
+    }
+    // Setup cleanup function, in case of failure below
+    auto fail_cleanup = [&topic_names_and_types]() {
+        rmw_ret_t rmw_ret = rmw_names_and_types_fini(topic_names_and_types);
+        if (rmw_ret != RMW_RET_OK) {
+          RCUTILS_LOG_ERROR_NAMED(
+            kLoggerTag,
+            "error during report of error: %s", rmw_get_error_string());
+        }
+      };
+    // Setup demangling functions based on no_demangle option
+    auto demangle_topic = _demangle_if_ros_topic;
+    auto demangle_type = _demangle_if_ros_type;
+    if (no_demangle) {
+      auto noop = [](const std::__cxx11::string & in) {
```

Comment:
It looks like macOS is cranky about this use of `std::__cxx11`: https://ci.ros2.org/job/ci_osx/4759/consoleFull, search for "__cxx11".  Can we just use `std::string` here?

Reply:
I went ahead and did it in 9ea473a .  If this isn't appropriate for some reason, we'll have to figure out another cross-platform way.

Reply:
Awesome, I went to change this and found it already there :)